Add kit_pse84_eval basic board and functionality support

mcatee-infineon opened the pull request:
- add dts, soc, and board files for ifx kit_pse84_eval board
- update pinctrl, clock_control, and uart drivers for kit_pse84_eval board
- update hal_infineon module for kit_pse84_eval board

nordicjm reviewed (changes requested):
this PR is wholly non zephyr coding style compliant

nordicjm commented on soc/infineon/edge/pse84/security_config/pse84_s_mpc.c:
```diff
@@ -0,0 +1,113 @@
+	return return_result;
+}
+
+cy_rslt_t cy_mpc_Init(void)
```
this is still not snake case

mcatee-infineon replied:
This should be resolved now.

nordicjm commented on modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt:
```diff
@@ -32,43 +41,68 @@ zephyr_include_directories_ifdef(CONFIG_SOC_FAMILY_INFINEON_CAT1C ${pdl_dev_cat1
 zephyr_library_sources_ifdef(CONFIG_USE_INFINEON_UART     ${pdl_drv_dir}/source/cy_scb_uart.c)
-zephyr_library_sources_ifdef(CONFIG_USE_INFINEON_FLASH    ${pdl_drv_dir}/source/cy_flash.c)
+if(CONFIG_SOC_FAMILY_INFINEON_EDGE)
+zephyr_code_relocate(FILES ${pdl_drv_dir}/source/cy_scb_uart.c LOCATION RAM)
```
not indented

mcatee-infineon replied:
This should be resolved now.

nordicjm commented on modules/hal_infineon/zephyr-ifx-cycfg/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_SOC_FAMILY_INFINEON_EDGE)
+set(zephyr_ifx_cycfg_dir ${ZEPHYR_HAL_INFINEON_MODULE_DIR}/zephyr-ifx-cycfg/kit_pse84_eval)
```
etc.

mcatee-infineon replied:
This should be resolved, with above.

nordicjm commented on modules/hal_infineon/mtb-template-cat1/CMakeLists.txt:
```diff
@@ -33,3 +40,21 @@ if(CONFIG_SOC_FAMILY_INFINEON_CAT1C)
+zephyr_include_directories(${edge_dir}/devices/include)
+
+zephyr_library_sources(${edge_dir}/system_edge.c)
+	if(CONFIG_BOARD_KIT_PSE84_EVAL_PSE846GPS2DBZC4A_M33)
```
no tabs in cmake

mcatee-infineon replied:
This should be fixed with a move to the 2 space indentation

nordicjm commented on boards/infineon/kit_pse84_eval/doc/kit_pse84_eval.webp:
this is still far too large

nordicjm commented on boards/infineon/kit_pse84_eval/kit_pse84_eval_common.dtsi:
```diff
@@ -0,0 +1,194 @@
+
+	leds {
+		compatible = "gpio-leds";
+		led_red: led_0 {
```
missing newlines e.g. line 19, 23, fix in whole PR

mcatee-infineon replied:
This should be fixed in the *_common.dtsi file. I did not see any other apparent instances, that had not been addressed yet.

nordicjm commented on boards/infineon/kit_pse84_eval/sysbuild.cmake:
```diff
@@ -0,0 +1,33 @@
+
+if(SB_CONFIG_BOARD_KIT_PSE84_EVAL_PSE846GPS2DBZC4A_M33)
+
+ExternalZephyrProject_Add(
```
@tomi-font

mcatee-infineon replied:
@nordicjm @tomi-font. 

The current implementation of the flow for the board's cores is to move from
m33_s -> m33_ns -> m55, in general samples and tests.

We created the sysbuild.cmake file to allow for the building of dependencies. i.e. m55 needs m33_ns/m33_s and m33_ns needs m33_s. The idea we had was to allow for the building and programming of each individual target, if desired. Meanwhile, if someone wanted to build a target with its "dependencies", they could use the "--sysbuild" flag. The intention is to use our specific secure application for the m33_s. 

This is similar to the "Building the Secure firmware using Zephyr" section of the index.rst for the nrf5340dk board, without the merging of binaries. We are using the CONFIG_TRUSTED_EXECUTION_SECURE and CONFIG_TRUSTED_EXECUTION_NONSECURE to differentiate between the m33_s and m33_ns. However, we are not using CONFIG_BUILD_WITH_TFM.

nordicjm commented on drivers/clock_control/clock_control_ifx_fixed_clock.c:
```diff
@@ -14,46 +14,109 @@
+	uint32_t system_clock; /* ifx_cat1_clock_block */
 };
 
+__WEAK void cycfg_ClockStartupError(uint32_t error)
```
as above, these are not snake case

mcatee-infineon commented on soc/infineon/edge/pse84/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+  zephyr_compile_definitions(FLASH_BOOT)
+
+  set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+      COMMAND edgeprotecttools image-metadata --image ${PROJECT_BINARY_DIR}/zephyr.hex  --output ${PROJECT_BINARY_DIR}/zephyr.hex --erased-val 0xff --header-size 0x400 --slot-size 0x130000  --hex-addr 0x60100000
```
This is not fixed in the current update, but will be addressed in the next update with something like a zephyr-signed.hex. Would that work?

nordicjm replied:
you need to get the bin name correctly since that is a Kconfig then you can set it to that + `.signed.hex` yes

sreeramIfx commented on modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt:
```diff
@@ -3,22 +3,31 @@
-set(pdl_dev_cat1b_dir   ${ZEPHYR_HAL_INFINEON_MODULE_DIR}/mtb-pdl-cat1/devices/COMPONENT_CAT1B)
-set(pdl_dev_cat1c_dir   ${ZEPHYR_HAL_INFINEON_MODULE_DIR}/mtb-pdl-cat1/devices/COMPONENT_CAT1C)
+if(CONFIG_SOC_FAMILY_INFINEON_EDGE)
+  set(pdl_dir ${ZEPHYR_HAL_INFINEON_MODULE_DIR}/mtb-dsl-pse8xxgp/pdl)
```
please create a separate cmakefile for mtb-dsl-pse8xxgp

mcatee-infineon replied:
The new cmakefile now exists.

mcatee-infineon commented on boards/infineon/kit_pse84_eval/kit_pse84_eval_m33.dts:
```diff
@@ -0,0 +1,37 @@
+	};
+};
+
+&clk_iho {
```
This should be fixed now.

mcatee-infineon commented on boards/infineon/kit_pse84_eval/kit_pse84_eval_m33.yaml:
```diff
@@ -0,0 +1,18 @@
+flash: 512
+toolchain:
+  - zephyr
+supported:
```
This is now added.

mcatee-infineon commented on soc/infineon/edge/pse84/security_config/pse84_boot.c:
```diff
@@ -0,0 +1,96 @@
+
+#include "pse84_boot.h"
+
+#define CYMEM_CM33_0_m33_nvm_START 0x60300000u
```
This should be fixed now.

tomi-font commented on soc/infineon/edge/pse84/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	  This option enables the m55 core on pse84. It is enabled during
+	  initialization of the cm33 core in its nonsecure state.
+
+config SOC_PSE84_M33_NS_ENABLE
```
Yeah, the one thing I'm wondering about is, what is done with the S/NS split here? There's no TF-M involved here, so @mcatee-infineon what is the goal with this?

mcatee-infineon replied:
The idea was to build and flash the S/NS independently. There are more details in the comment on the related thread. 
https://github.com/zephyrproject-rtos/zephyr/pull/95874#discussion_r2361358322

mcatee-infineon commented on modules/hal_infineon/CMakeLists.txt:
```diff
@@ -51,10 +66,10 @@ if (CONFIG_WIFI_AIROC)
-if (CONFIG_BT_PSOC6_BLESS)
-add_subdirectory(bless)
-endif()
+  if(CONFIG_BT_PSOC6_BLESS)
```
As above

mcatee-infineon commented on modules/hal_infineon/mtb-dsl-pse8xxgp/CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+endif()
+
+# Peripheral drivers
+zephyr_library_sources_ifdef(CONFIG_USE_INFINEON_I2C      ${pdl_drv_dir}/source/cy_scb_i2c.c)
```
This section should now have the extra spaces removed.

mcatee-infineon commented on dts/arm/infineon/cat1b/psc3/system_clocks.dtsi:
```diff
@@ -99,310 +97,311 @@
-			div-type = <DIV_8_BIT>;
-			div-num = <3>;
-			div-value = <1>;
+				#clock-cells = <0>;
```
This should now be fixed.

mcatee-infineon committed:
soc: add support for ifx edge socs

- add basic soc files to support ifx pse84 soc
- add files needed for setting up the m55

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon committed:
dts: add devicetree support for ifx edge devices

- add devicetree support for pse84 devices

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon committed:
boards: add support for ifx kit_pse84_eval

- add needed board files for kit_pse84_eval board

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon committed:
drivers: clock_control: rename files related to ifx clock_control

- rename ifx clock_control drivers
- rename infineon,cat1-peri-div yaml
- rename ifx clock dt-binding .h file

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon committed:
drivers: clock_control: add clock_control support for kit_pse84_eval

- add support for kit_pse84_eval board
- refactor infineon,fixed-clock binding
- refactor infineon,fixed-factor binding
- refactor infinein,peri-div binding

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon committed:
drivers: uart: update ifx uart_pdl driver for kit_pse84_eval

- update ifx uart driver for kit_pse84_eval board

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon committed:
drivers: add ifx pinctrl driver updates for kit_pse84_eval

- add drive-strength capability for kit_pse84_eval

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon committed:
tests: add kit_pse84_eval overlays for gpio_basic_api testsuite

- add kit_pse84_eval overlays for gpio_basic_api testsuite

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

nordicjm commented on modules/hal_infineon/CMakeLists.txt:
```diff
@@ -2,28 +2,30 @@
 # Copyright (c) 2022 Cypress Semiconductor Corporation.
 # SPDX-License-Identifier: Apache-2.0
 
-if(CONFIG_HAS_XMCLIB OR CONFIG_SOC_FAMILY_PSOC6_LEGACY OR CONFIG_SOC_FAMILY_INFINEON_CAT1)
```
also does this board build if you go to commit `2920e1779907b177c75b9d7a65aadaf927dcf5f4` and build it i.e. without the HAL update or other commits? If so, ignore message, if not, reorder commits with proper dependencies

mcatee-infineon replied:
The hal_infineon update commit has been reordered as the first commit in this PR.

mcatee-infineon commented on soc/infineon/edge/pse84/soc_pse84_m33_s.c:
```diff
@@ -0,0 +1,139 @@
+
+#define CY_IPC_MAX_ENDPOINTS (8UL)
+
+cy_en_sysint_status_t Cy_SysInt_Init(const cy_stc_sysint_t *config, cy_israddress userIsr)
```
This should be fixed now.

mcatee-infineon commented on soc/infineon/edge/pse84/soc_pse84_m55.c:
```diff
@@ -0,0 +1,148 @@
+	ARM_MPU_Enable(4);
+}
+
+cy_en_sysint_status_t Cy_SysInt_Init(const cy_stc_sysint_t *config, cy_israddress userIsr)
```
As above

mcatee-infineon commented on modules/hal_infineon/mtb-template-cat1/CMakeLists.txt:
```diff
@@ -33,3 +40,21 @@ if(CONFIG_SOC_FAMILY_INFINEON_CAT1C)
 endif()
+
+# Add support for psoce84 (EDGE)
+if (CONFIG_SOC_FAMILY_INFINEON_EDGE)
```
This should be fixed.

tejlmand commented on modules/hal_infineon/mtb-dsl-pse8xxgp/CMakeLists.txt:
```diff
@@ -0,0 +1,103 @@
+set(pdl_dev_edge_dir ${pdl_dir}/devices)
+
+# Generate PDL specific SOC defines
+zephyr_compile_definitions($<UPPER_CASE:${CONFIG_SOC}>)
```
but why is the high-level Zephyr library `zephyr` having source files which includes hal_infineon specific headers ?

Afaict, this CMake file lacks a `zephyr_library()` call at the top of the file.

tejlmand replied:
> The reason these defines (COMPONENT_SECURE_DEVICE & PSE846GPS2DBZC4A) are set, is in order for the correct header file to be aliased for and included by the soc during builds.

I fully understand, but the more problematic parts on such approach are:
- Flags are applied to all source files, even to those not requiring the flag.
   This results in unnecessary rebuilds of source files during incremental builds if a flag is changed, compared to having the flag on a specific library. Having a flag on a specific library will only rebuild source files in this lib when flag changes.
   A single flag related to the SoC itself is usually ok, but we should be careful not to apply unrelated compile flags to all sources just because it's convenient.
- The compile flag may be applied anywhere, making it hard to trace back the origin of the flag.
   In comparison, a mapping / config header file like used by mbedTLS or here: https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/nrfx/nrfx_kconfig.h are at a central location and comes without extra indirect dependencies.
   Having such a file also ensures a single location for looking up such mapping (see also next bullet).
- Kconfig to compile definition mapping may happen anywhere in CMake code, thus create hidden dependency adjustment which are hard to trace back. For example, I can directly seet that `COMPONENT_SECURE_DEVICE` are directly guarded by `CONFIG_SOC_FAMILY_INFINEON_EDGE`, `CONFIG_CPU_CORTEX_M33`, `CONFIG_TRUSTED_EXECUTION_SECURE`, and because this happens in the Zephyr module's top-level CMakeList glue file, then I probably can stop there, but a compile def placed elsewhere could have other dependencies deciding if the flag gets included or not, for example because code was only icluded with `add_subdirectory_ifdef(BAR <include>)` would create an indirect dep to BAR.
   If all dependency handling instead is ensured in Kconfig, then dependencies are much cleaner handled and we have tools to check for things like circular deps as well as finding all deps for a symbol easily.
- Using CMAke may result in duplication of flags.
   In this case there is a dep on `CONFIG_CPU_CORTEX_M33`, but someone requiring same flag elsewhere could end-up having `zephyr_compile_definitions_ifdef(FOO COMPONENT_SECURE_DEVICE)`
   Again, using Kconfig -> header mapping would help and allow to use the power of Kconfig and not having to do mapping in CMake.
- Long command line invocation may cause CMake to create a shell script for compiler invocation (bat-file on Windows).
   This further makes debugging / tracing hard, as a `ninja -v` will just show bat file invocation but not the content of said file on Windows. (though technically also a risk of Linux / MacOS, then this in practice mostly a Windows issue)

The above are some of the reasons why we should be careful on the approach and try to limit to very few flags, such as CPU type (if really needed), but not all sorts of compile definitions.

> It seems like something similar is being done by other vendors, as well.

We are aware, and the Nordic developers has been requested to cleanup the parts not related to the SoC itself for reasons given above. But it's always harder to  get people to cleanup afterwards than to catch things early when they are in a PR, and before it grows.
One reason we have started to pay a bit more attention to this.

> The zephyr_library_named() is being defined at the top level for hal_infineon.

Thanks, that was why I didn't spot it at first :+1: 

> I am not sure if I fully understand the question.

Hope this makes it a bit clearer, and feel free to follow-up with more questions.

mcatee-infineon committed:
modules: update revision of hal_infineon module

- update hal_infineon revision in preparation for new kit_pse84_eval
  board

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon committed:
modules: update hal_infineon module for infineon "edge" soc

- update moudle CMakeLists.txt files for new assets
- add infineon_kconfig.h file for definitions needed by soc

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

nordicjm reviewed (approved):
Looks good, thanks for fixing issues
